pkg/kube/namespaces.go

```diff
@@ -0,0 +1,218 @@
+package kube
+
+import (
+	"context"
+	"net"
+	"os"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/rancher/prometheus-auth/pkg/data"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	"golang.org/x/sync/singleflight"
+	authorization "k8s.io/api/authorization/v1"
+	core "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	authorizationTyped "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+type Namespaces interface {
+	Query(token string) data.Set
+}
+
+type namespaces struct {
+	singleflight.Group
+
+	ctx              context.Context
+	secretIndexer    cache.Indexer
+	namespaceIndexer cache.Indexer
+}
+
+func (n *namespaces) Query(token string) data.Set {
+	ret, err, _ := n.Do(token, func() (interface{}, error) {
```

Comment:
while singleflight is a pretty cool approach I don't think it does really the right thing.  We want to call SubjectAccessReview and if true cache the true result for say 5 minutes.  If false don't cache.  So that way we only make 1 SAR request per token every 5 minutes. singleflight will only consolidate concurrent requests, not cache.

Reply:
Fixed, using `k8s.io/apimachinery/pkg/util/cache`.